ctk/common/src/main/kotlin/org/codice/compliance/verification/core/ResponseProtocolVerifier.kt

```diff
@@ -15,155 +15,102 @@ package org.codice.compliance.verification.core
-                    parent = RESPONSE,
-                    node = response)
+        CommonDataTypeVerifier.verifyStringValues(version)
         CommonDataTypeVerifier.verifyTimeValues(
-                response.attributes.getNamedItem(ISSUE_INSTANT), SAMLCore_3_2_2_d)
+                response.attributes.getNamedItem("IssueInstant"), SAMLCore_3_2_2_d)
 
-        val destination = response.attributes?.getNamedItem("Destination")?.textContent
-        if (destination != acsUrl)
-            throw SAMLComplianceException.createWithPropertyMessage(SAMLCore_3_2_2_e,
-                    property = "Destination",
-                    actual = destination,
-                    expected = acsUrl ?: "No ACS URL Found",
-                    node = response)
+        val destination = response.attributes?.getNamedItem(DESTINATION)
+        if (destination != null) {
+            if (destination.textContent != acsUrl)
+                throw SAMLComplianceException.createWithPropertyMessage(SAMLCore_3_2_2_e,
+                        property = DESTINATION,
+                        actual = destination.textContent,
+                        expected = acsUrl ?: "No ACS URL Found",
+                        node = response)
 
-        if (response.children(STATUS).isEmpty())
-            throw SAMLComplianceException.createWithXmlPropertyReqMessage(SAMLCore_3_2_2,
-                    property = STATUS,
-                    parent = RESPONSE,
-                    node = response)
+            CommonDataTypeVerifier.verifyUriValues(destination)
+        }
+
+        val content = response.attributes?.getNamedItem(DESTINATION)
+        if (content != null) {
+            CommonDataTypeVerifier.verifyUriValues(content)
+        }
     }
 
-    /**
-     * Verify the Statuses and Status Codes
-     * 3.2.2.1 Element <Status>
-     * 3.2.2.2 Element <StatusCode>
-     */
+    /** 3.2.2.2 Element <StatusCode> */
     private fun verifyStatusesType() {
-        // Status
-        response.children(STATUS).forEach {
-            val statusCodes = it.children(STATUS_CODE)
-            if (statusCodes.isEmpty())
-                throw SAMLComplianceException.createWithXmlPropertyReqMessage("SAMLCore.3.2.2.1",
-                        property = STATUS_CODE,
-                        parent = STATUS,
-                        node = response)
-
-            // StatusCode
-            if (statusCodes.any { it.attributes.getNamedItem(VALUE) == null })
-                throw SAMLComplianceException.createWithXmlPropertyReqMessage("SAMLCore.3.2.2.2",
-                        property = VALUE,
-                        parent = STATUS_CODE,
-                        node = response)
+        if (response.allChildren("Status")
+                        .any {
+                            !TOP_LEVEL_STATUS_CODES
+                                    .contains(it.children(STATUS_CODE).first()
+                                            .attributes.getNamedItem("Value").textContent)
```

Comment:
❗️ The StatusCode is optional, so we may get a NoSuchElementException calling first blindly

Reply:
`<StatusCode>` is a required element for `<Status>`s (Core 3.2.2.1)

Reply:
Ah, sorry. Was looking at `3.2.2.2`